Refactor `LibGit2.set_remote_url`

omus opened the pull request:
Added new functions `set_remote_fetch_url` and `set_remote_push_url` which allow you to specify the fetch/push remote URLs independently. Additionally, I switched `set_remote_url` from using a the keyword remote to using an optional argument.

Fixes https://github.com/JuliaLang/julia/issues/21773

omus committed:
Allow setting the remote push/fetch URL separately

Changes include:
- Creating `set_remote_fetch_url`, `set_remote_push_url`,
  `remote_delete`, and `lookup_remote`
- Updating `set_remote_url` to use LibGit2 calls
- Deprecate `set_remote_url` which uses keyword and reorder arguments to
  match order of `GitRemote` constructor.
- Update all existing calls to `set_remote_url`

omus commented on base/libgit2/libgit2.jl:
```diff
@@ -240,49 +240,6 @@ function is_ancestor_of(a::AbstractString, b::AbstractString, repo::GitRepo)
-            if push != url
-                set!(cfg, "remote.$remote.pushurl", push)
-            end
-        end
```
Caused #21773

StefanKarpinski commented:
Feature or bugfix?

tkelman commented:
looks like a little of both, plus a new deprecation

tkelman commented on base/libgit2/remote.jl:
```diff
@@ -241,3 +265,69 @@ function push(rmt::GitRemote, refspecs::Vector{<:AbstractString};
+repo_path = joinpath("test_directory", "Example")
+repo = LibGit2.init(repo_path)
+url1 = "https://github.com/JuliaLang/Example.jl"
+LibGit2.set_remote_url(repo, url1, remote="upstream")
```
bit misleading to name the argument at the call site if it's positional

omus replied:
I missed this when I moved away from a keyword.

omus replied:
Fixed

tkelman commented on base/deprecated.jl:
```diff
@@ -1348,6 +1348,18 @@ end
+    if remote_kwarg === nothing
+        return remote_arg
+    else
+        msg = "$f($sig; remote=\"$remote_kwarg\") is deprecated, use $f($sig, \"$remote_kwarg\")"
```
these usually end with "instead"

omus replied:
Ok. I modeled this off of the `deprecate_negate` function. Does the rest of this seem reasonable?

tkelman replied:
I wrote that one, and went through a couple refactors of it. it does say instead in it.

the argument order differs from the libgit2 c functions', but I guess that's okay. positionals always run the risk of "what order is it supposed to be" when writing or reading code that uses the method, so I'm neutral on this part of the change

omus replied:
> I modeled this off of the deprecate_negate function. Does the rest of this seem reasonable?

I should have been more specific. I was wondering if the `deprecate_remote_keyword` function was reasonable.

omus replied:
> the argument order differs from the libgit2 c functions'

Re-ordering the functions to be `func(repo::GitRepo, remote::AbstractString, url::AbstractString)` would more closely match both LibGit2 and the git command line. Personally I am fine with always having to specify the remote name.

omus replied:
Done

tkelman commented on test/libgit2.jl:
```diff
@@ -37,6 +37,15 @@ function with_libgit2_temp_home(f)
 
+function is_defined_remote(repo::LibGit2.GitRepo, remote::AbstractString)
+    try
+        LibGit2.get(LibGit2.GitRemote, repo, name)
```
can this be done without try-catching?

omus replied:
I'll look into this again tomorrow. I didn't find a LibGit2 method which seemed to cover this.

omus replied:
Made `lookup_remote` to address this

tkelman commented on test/libgit2.jl:
```diff
@@ -1073,6 +1082,47 @@ mktempdir() do dir
+            @test !is_defined_remote(repo, name)
+
+            # Set just the fetch URL
+            LibGit2.set_remote_fetch_url(repo, url, name)
```
this is testing the `repo::GitRepo` versions pretty well, and the existing block was testing some of the `path::AbstractString` versions, but not the ones you're newly adding

omus replied:
I'll extend the tests to cover those.

omus replied:
Done

omus commented:
> Feature or bugfix?

Definitely both. When addressing the bug I noticed that the GitConfig was being used to modify the remote rather than using the available LibGit2 functions. Using the config method is bad as it doesn't verify that things like the remote name is valid. After getting my hands into the code I noticed that the remote keyword was unnecessary which is how I ended up here.

omus commented:
I'll add some docstrings to the new functions in addition to the review change requests.

StefanKarpinski commented:
I guess what I was getting at was is this backportable to 0.6 or does it need to stay on master?

omus commented:
> I guess what I was getting at was is this backportable to 0.6 or does it need to stay on master?

I can separate the bugfix so it can be backported to 0.6.

omus committed:
Fix issue #21773

When using `LibGit2.set_remote_url` to specify a GitHub HTTP URL only
the fetch URL would be modified to use HTTP and the push URL would
always be set to SSH.

omus commented:
A note for the future @tkelman: Only backport commit ed263e8

omus commented:
Anything left to do here? Maybe a NEWS entry?

tkelman commented on base/libgit2/remote.jl:
```diff
@@ -240,4 +285,89 @@ function push(rmt::GitRemote, refspecs::Vector{<:AbstractString};
+function set_remote_push_url end
+
+function set_remote_push_url(repo::GitRepo, remote_name::AbstractString, url::AbstractString)
+@check ccall((:git_remote_set_pushurl, :libgit2), Cint,
```
under-indented

omus replied:
Nice catch

omus commented:
Rebased to resolve conflict in "base/deprecated.jl"

omus commented:
Ready for merge

tkelman commented:
didn't add a news entry, were you going to?

omus commented:
I wasn't planning on adding one but I will now.

omus committed:
Add NEWS entry

[ci skip]

omus commented:
NEWS entry added.
